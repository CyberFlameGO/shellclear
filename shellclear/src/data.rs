use crate::shell::Shell;
use serde_derive::Deserialize;

pub struct CmdExit {
    pub code: exitcode::ExitCode,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct SensitiveCommands {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub name: String,
}

#[derive(Debug)]
pub struct FindingSensitiveCommands {
    pub shell_type: Shell,
    pub sensitive_findings: Vec<SensitiveCommands>,
    pub command: String,
}
