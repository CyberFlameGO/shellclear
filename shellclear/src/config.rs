use crate::data::{SensitiveCommands, ROOT_APP_FOLDER};
use anyhow::Result;
use std::fs;
use std::path::PathBuf;

const CONFIG_SENSITIVE_PATTERNS: &str = "sensitive-patterns.yaml";
const SENSITIVE_PATTERN_TEMPLATE: &str = r###"# External sensitive patters file allows you you add a custom patterns to shellclear

- name: Pattern Name
  test: <PATTERN REGEX>
"###;

#[derive(Clone, Debug)]
pub struct Config {
    sensitive_commands_path: PathBuf,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            sensitive_commands_path: Config::get_sensitive_pattern_file(&dirs::home_dir().unwrap()),
        }
    }
}

impl Config {
    fn with_custom_path(root: &PathBuf) -> Self {
        Self {
            sensitive_commands_path: Self::get_sensitive_pattern_file(root),
        }
    }
    /// Init external configuration
    ///
    /// # Errors
    ///
    /// Will return `Err` when home directory not found or failed to create a file
    pub fn init(&self) -> Result<String> {
        let file = &self.sensitive_commands_path;
        fs::write(file, SENSITIVE_PATTERN_TEMPLATE)?;
        Ok(file.display().to_string())
    }

    /// Get sensitive file path name
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found
    #[must_use]
    pub fn get_sensitive_pattern_name(&self) -> String {
        self.sensitive_commands_path.display().to_string()
    }

    /// Is sensitive file is exists
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found
    #[must_use]
    pub fn is_sensitive_pattern_file_exists(&self) -> bool {
        self.sensitive_commands_path.exists()
    }

    /// Load sensitive pattern file
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found or yaml is invalid
    pub fn load_patterns_from_default_path(&self) -> Result<Vec<SensitiveCommands>> {
        self.load_patterns_from_file(&self.sensitive_commands_path)
    }

    /// Load sensitive pattern from the given path
    ///
    /// # Errors
    ///
    /// Will return `Err`  yaml is invalid
    fn load_patterns_from_file(&self, path: &PathBuf) -> Result<Vec<SensitiveCommands>> {
        let f = std::fs::File::open(path)?;
        Ok(serde_yaml::from_reader(f)?)
    }

    /// Returns the root shellclear config folder
    fn get_sensitive_pattern_file(path: &PathBuf) -> PathBuf {
        path.join(format!(".{}", ROOT_APP_FOLDER))
            .join(CONFIG_SENSITIVE_PATTERNS)
    }
}
