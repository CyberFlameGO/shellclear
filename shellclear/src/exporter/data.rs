use crate::data::FindingSensitiveCommands;
use crate::shell::Shell;
use anyhow::Result;
use chrono::{DateTime, Local, NaiveDateTime, TimeZone};
use lazy_static::lazy_static;
use regex::Regex;

const DATE_TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";

pub trait Exporter {
    fn sensitive_data(&self, findings: &[&FindingSensitiveCommands]) -> Result<()>;
}

lazy_static! {
    static ref ZSHRC_CAPTURE_TIME: Regex = Regex::new(r"^: ([0-9]+):").unwrap();
}

// todo:: move under export trait
pub fn chunk(text: &str, size: usize) -> String {
    text.chars()
        .collect::<Vec<char>>()
        .chunks(size)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<String>>()
        .join("\r\n")
}

/// Get command time execution
///
/// # Errors
///
/// Will return `Err` when conversion error
pub fn extract_time(finding: &FindingSensitiveCommands) -> Result<String> {
    match finding.shell_type {
        Shell::Zshrc => {
            if let Some(c) = ZSHRC_CAPTURE_TIME.captures(&finding.data) {
                if let Some(timestemp) = c.get(1) {
                    let date_time: DateTime<Local> = Local.from_utc_datetime(
                        &NaiveDateTime::from_timestamp(timestemp.as_str().parse::<i64>()?, 0),
                    );
                    return Ok(format!("{}", date_time.format(DATE_TIME_FORMAT)));
                }
            };
            Ok(finding.data.clone())
        }
        _ => Ok("".to_string()),
    }
}
