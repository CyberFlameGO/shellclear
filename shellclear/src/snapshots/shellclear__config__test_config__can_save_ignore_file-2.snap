---
source: shellclear/src/config.rs
expression: config.get_ignore_patterns()
---
Ok(
    [
        "patter-1",
        "patter-2",
    ],
)
